docs/architecture/adr-030-msg-auth-module.md

```diff
@@ -0,0 +1,222 @@
+the `Msg` method is defined. `Authorization`s use the new `ServiceMsg` type from
+ADR 031.
+
+#### Authorization
+
+```go
+type Authorization interface {
+	// MethodName returns the fully-qualified Msg service method name as described in ADR 031.
+	MethodName() string
+
+	// Accept determines whether this grant permits the provided sdk.ServiceMsg to be performed, and if
+	// so provides an upgraded authorization instance.
+	Accept(msg sdk.ServiceMsg, block abci.Header) (allow bool, updated Authorization, delete bool)
+}
+```
+
+For example a `SendAuthorization` like this is defined for `MsgSend` that takes
+a `SpendLimit` and updates it down to zero:
+
+```go
+type SendAuthorization struct {
+	// SpendLimit specifies the maximum amount of tokens that can be spent
+	// by this authorization and will be updated as tokens are spent. If it is
+	// empty, there is no spend limit and any amount of coins can be spent.
+	SpendLimit sdk.Coins
+}
+
+func (cap SendAuthorization) MethodName() string {
+	return "/cosmos.bank.v1beta1.Msg/Send"
+}
+
+func (cap SendAuthorization) Accept(msg sdk.ServiceMsg, block abci.Header) (allow bool, updated Authorization, delete bool) {
+	switch req := msg.Request.(type) {
+	case bank.MsgSend:
+		left, invalid := cap.SpendLimit.SafeSub(req.Amount)
+		if invalid {
+			return false, nil, false
+		}
+		if left.IsZero() {
+			return true, nil, true
+		}
+		return true, SendAuthorization{SpendLimit: left}, false
+	}
+	return false, nil, false
+}
+```
+
+A different type of capability for `MsgSend` could be implemented
+using the `Authorization` interface with no need to change the underlying
+`bank` module.
+
+### `Msg` Service
+
+```proto
+service Msg {
+  // GrantAuthorization grants the provided authorization to the grantee on the granter's
+  // account with the provided expiration time.
+  rpc GrantAuthorization(MsgGrantAuthorization) returns (MsgGrantAuthorizationResponse);
+
+  // ExecAuthorized attempts to execute the provided messages using
+  // authorizations granted to the grantee. Each message should have only
+  // one signer corresponding to the granter of the authorization.
+  rpc ExecAuthorized(MsgExecAuthorized) returns (MsgExecAuthorizedResponse)
+
+
+  // RevokeAuthorization revokes any authorization corresponding to the provided method name on the
+  // granter's account that has been granted to the grantee.
+  rpc RevokeAuthorization(MsgRevokeAuthorization) returns (MsgRevokeAuthorizationResponse);
+}
+
+message MsgGrantAuthorization{
+  string granter = 1;
+  string grantee = 2;
+  google.protobuf.Any authorization = 3 [(cosmos_proto.accepts_interface) = "Authorization"];
+  google.protobuf.Timestamp expiration = 4;
+}
+
+message MsgExecAuthorized {
+    string grantee = 1;
+    repeated google.protobuf.Any msgs = 2;
+}
+
+message MsgRevokeAuthorization{
+  string granter = 1;
+  string grantee = 2;
+  string method_name = 3;
+}
```

Comment:
I'm thinking loudly: this means that we can only register one rule for one message for one grantee. That is defined in through the `authorization` attribute of `MsgGrantAuthorization`. Let's consider this example, Alice wants to grant an access to Bob with following conditions:
* Bob can spend max 100 USD per month. This authorization expires in 2020
* Bob can spend max 200 USD per month in 2021

To do that we would need an authorization ID generated and returned in `GrantAuthorization`

Reply:
One grant per granter/grantee/msg-type. So it's unique for those three. The above case you described could be handled by some custom `Authorization` type - maybe `AuthorizationGroup` or something that Alice would configuration.

Reply:
We could allow Alice to grant Bob multiple spend authorizations like you suggest, but is that necessary?

Reply:
If this complicates, then it's not necessary. App developer can always design more complex Authorization structure to handle `or` conditions.

Reply:
> App developer can always design more complex Authorization structure to handle or conditions.

I think a nice library of composable authorization structures would be nice.

Reply:
Let's follow up on this in https://github.com/cosmos/cosmos-sdk/issues/8312